### PR TITLE
Use fuzzing to check invariants

### PR DESCRIPTION
**Pull request by markspanbroek**
Simplifies the contract code by moving the invariant checks to separate files.
Uses [echidna](https://github.com/crytic/echidna) to perform the fuzzing.

### CODE REVIEW

File: fuzzing/fuzz.sh [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+else
+  fuzz () {
+    docker run \
+      --rm \
+      -v ${root}:/src ghcr.io/crytic/echidna/echidna \
+      bash -c \
+        "cd /src && echidna-test . \
+          --config fuzzing/echidna.yaml \
+          --corpus-dir fuzzing/corpus \
+          --crytic-args --ignore-compile \
```
Comment (emizzle) [inline fuzzing/fuzz.sh:22]: Noticed that `--crytic-args --ignore-compile` flags were not in the non-docker command. Just checking that these weren't added inadvertently.

File: contracts/Collateral.sol [hunk 2]
```diff
@@ -5,71 +5,46 @@ import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
 
 abstract contract Collateral {
   IERC20 public immutable token;
-  CollateralFunds private _funds;
+  CollateralTotals internal _collateralTotals;
 
   mapping(address => uint256) private _balances;
 
-  constructor(IERC20 token_) collateralInvariant {
```
Comment (emizzle) [inline contracts/Collateral.sol:12]: I'm likely missing something, but can we assume that the inputs tested by the fuzzer and the passing result means that we no longer need to enforce parameters using an invariant? Or is there something I am missing where this invariant is allowed to be removed but somehow is still enforced?

**Review by emizzle — APPROVED** [review]
Nice work, Mark! Looks good, and I can see it ran on CI successfully which is awesome. I have mainly one question in the code, and also I was having trouble running this locally if you would mind taking a look:

Locally, i ran into some issues running this branch, in chronological order:
1. Docker wasn't running (error output = `docker: Error response from daemon: dial unix /Users/emizzle/Library/Containers/com.docker.docker/Data/docker.raw.sock: connect: connection refused.`) Solved by running Docker 😄 
2. I wasn't logged into docker (error output = `Unable to find image 'ghcr.io/crytic/echidna/echidna:latest' locally
docker: Error response from daemon: Bad response from Docker engine.`). Solved by logging into Docker.
3. Could not get past:
```bash
❯ npm run fuzz

> fuzz
> fuzzing/fuzz.sh

Unable to find image 'ghcr.io/crytic/echidna/echidna:latest' locally
latest: Pulling from crytic/echidna/echidna
846c0b181fff: Pull complete 
a29861e18734: Pull complete 
929040e99d59: Pull complete 
33e42817120e: Pull complete 
8a58b4a70271: Pull complete 
Digest: sha256:82042765e513e3c5979d3edb3b46b8718b25bc2b48ea8971be76537a8cac5f8d
Status: Downloaded newer image for ghcr.io/crytic/echidna/echidna:latest
Loaded total of 0 transactions from fuzzing/corpus/reproducers/
Loaded total of 0 transactions from fuzzing/corpus/coverage/ # <=== Diff from CI!
echidna-test: Given contract "FuzzCollateral" not found in given file
```
CI shows:
```bash
Run npm run fuzz

> fuzz
> fuzzing/fuzz.sh

Unable to find image 'ghcr.io/crytic/echidna/echidna:latest' locally
latest: Pulling from crytic/echidna/echidna
84[6](https://github.com/status-im/codex-contracts-eth/pull/37/checks#step:7:7)c0b181fff: Already exists
a29861e18[7](https://github.com/status-im/codex-contracts-eth/pull/37/checks#step:7:8)34: Pulling fs layer
929040e99d59: Pulling fs layer
33e42[8](https://github.com/status-im/codex-contracts-eth/pull/37/checks#step:7:9)17120e: Pulling fs layer
8a58b4a70271: Pulling fs layer
8a58b4a70271: Waiting
[9](https://github.com/status-im/codex-contracts-eth/pull/37/checks#step:7:10)29040e99d59: Verifying Checksum
929040e99d59: Download complete
33e42817[12](https://github.com/status-im/codex-contracts-eth/pull/37/checks#step:7:13)0e: Verifying Checksum
33e42817120e: Download complete
a29861e18734: Verifying Checksum
a29861e18734: Download complete
8a58b4a70271: Verifying Checksum
8a58b4a70271: Download complete
a29861e18734: Pull complete
929040e99d59: Pull complete
33e42817120e: Pull complete
8a58b4a70271: Pull complete
Digest: sha256:82042765e5[13](https://github.com/status-im/codex-contracts-eth/pull/37/checks#step:7:14)e3c5979d3edb3b46b87[18](https://github.com/status-im/codex-contracts-eth/pull/37/checks#step:7:19)b[25](https://github.com/status-im/codex-contracts-eth/pull/37/checks#step:7:26)bc2b48ea8971be76537a8cac5f8d
Status: Downloaded newer image for ghcr.io/crytic/echidna/echidna:latest
Loaded total of 0 transactions from fuzzing/corpus/reproducers/
Loaded total of 36 transactions from fuzzing/corpus/coverage/ # <=== Diff from local!
Analyzing contract: /src/contracts/FuzzCollateral.sol:FuzzCollateral
neverLetIndividualBalanceBeMoreThanAvailableFunds():  passed! 🎉
slash(address,uint256):  passed! 🎉
neverMoveSlashedFundsOut():  passed! 🎉
neverLoseFunds():  passed! 🎉
# ... removed for brevity
```
Somehow locally `fuzzing/corpus/coverage` isn't producing the transactions needed to run the fuzzer.

File: contracts/Collateral.sol (hunk 2 above)
Comment (markspanbroek) [reply contracts/Collateral.sol:12]: That's exactly it; by performing enough fuzzing we can gain confidence that the invariants cannot be invalidated, which negates the need for runtime checks. Ideally we would prove that the invariants cannot be invalidated, but that turned out to be problematic: #35

**Comment by markspanbroek** [comment]
> Somehow locally `fuzzing/corpus/coverage` isn't producing the transactions needed to run the fuzzer.

No need to worry about the corpus; this is a list of transactions that the fuzzer tried before, which it uses to be more efficient in finding cases that it hasn't covered before. It starts out empty, which is what you see on your development machine. I've configured the CI to cache the results, so that it remembers its corpus over multiple runs.

The real reason that your run fails, seems to be this line:
```
echidna-test: Given contract "FuzzCollateral" not found in given file
```

Can you try to compile the contracts beforehand? For instance using `npx hardhat compile`? If that solves the problem then we probably need to include that step in `fuzz.sh`, just before running docker.

**Comment by emizzle** [comment]
> > Somehow locally `fuzzing/corpus/coverage` isn't producing the transactions needed to run the fuzzer.
> 
> No need to worry about the corpus; this is a list of transactions that the fuzzer tried before, which it uses to be more efficient in finding cases that it hasn't covered before. It starts out empty, which is what you see on your development machine. I've configured the CI to cache the results, so that it remembers its corpus over multiple runs.
> 
> The real reason that your run fails, seems to be this line:
> 
> ```
> echidna-test: Given contract "FuzzCollateral" not found in given file
> ```
> 
> Can you try to compile the contracts beforehand? For instance using `npx hardhat compile`? If that solves the problem then we probably need to include that step in `fuzz.sh`, just before running docker.

I did try compiling the contracts earlier via `npm test`, however the artifacts may have been deleted after the run, so I compiled via `npx hardhat compile` then `npm run fuzz`, and got a "blank" error:
```
❯ npx hardhat compile
Compiled 6 Solidity files successfully
❯ npm run fuzz

> fuzz
> fuzzing/fuzz.sh

Loaded total of 0 transactions from fuzzing/corpus/reproducers/
Loaded total of 0 transactions from fuzzing/corpus/coverage/
Analyzing contract: /src/contracts/FuzzCollateral.sol:FuzzCollateral
echidna-test: Error running slither:

```

**Comment by markspanbroek** [comment]
Thanks for testing this earlier @emizzle! I have added a compile step to `npm run fuzz` to make sure that the contracts are compiled before invoking echidna. This makes the fuzzing work on a clean debian vm with only docker and nodejs installed. But I suspect it doesn't solve the problem that you're seeing on your system. Perhaps we can schedule a call to debug this further?

**Comment by emizzle** [comment]
Thanks for updating this @markspanbroek!

I have a new system since my last test unfortunately, and am running ARM64, so now I'm seeing the following error:
```shell
❯ npm run fuzz                    

> fuzz
> hardhat compile && fuzzing/fuzz.sh

Nothing to compile
Unable to find image 'ghcr.io/crytic/echidna/echidna:latest' locally
latest: Pulling from crytic/echidna/echidna
846c0b181fff: Pull complete 
a29861e18734: Pull complete 
929040e99d59: Pull complete 
33e42817120e: Pull complete 
8a58b4a70271: Pull complete 
Digest: sha256:82042765e513e3c5979d3edb3b46b8718b25bc2b48ea8971be76537a8cac5f8d
Status: Downloaded newer image for ghcr.io/crytic/echidna/echidna:latest
WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
Loaded total of 0 transactions from fuzzing/corpus/reproducers/
Loaded total of 0 transactions from fuzzing/corpus/coverage/
Analyzing contract: /src/contracts/FuzzCollateral.sol:FuzzCollateral
echidna-test: Error running slither:


```

**Comment by emizzle** [comment]
> Thanks for updating this @markspanbroek!
> 
> I have a new system since my last test unfortunately, and am running ARM64, so now I'm seeing the following error:
> 
> ```shell
> ❯ npm run fuzz                    
> 
> > fuzz
> > hardhat compile && fuzzing/fuzz.sh
> 
> Nothing to compile
> Unable to find image 'ghcr.io/crytic/echidna/echidna:latest' locally
> latest: Pulling from crytic/echidna/echidna
> 846c0b181fff: Pull complete 
> a29861e18734: Pull complete 
> 929040e99d59: Pull complete 
> 33e42817120e: Pull complete 
> 8a58b4a70271: Pull complete 
> Digest: sha256:82042765e513e3c5979d3edb3b46b8718b25bc2b48ea8971be76537a8cac5f8d
> Status: Downloaded newer image for ghcr.io/crytic/echidna/echidna:latest
> WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
> Loaded total of 0 transactions from fuzzing/corpus/reproducers/
> Loaded total of 0 transactions from fuzzing/corpus/coverage/
> Analyzing contract: /src/contracts/FuzzCollateral.sol:FuzzCollateral
> echidna-test: Error running slither:
> ```

Looks like echidna doesn't support arm docker images, though they could be built with emulation. We could contribute to echidna a cross-platform ci solution that builds both amd64 and arm64 images.

**Comment by markspanbroek** [comment]
> Looks like echidna doesn't support arm docker images

I added a check to the fuzz.sh script, so that it gives a more useful error message when running on arm systems.

File: fuzzing/fuzz.sh [hunk 3]
```diff
@@ -0,0 +1,33 @@
+#!/bin/bash
```
Comment (AuHau) [inline fuzzing/fuzz.sh:1]: Generally you should use this ;-)
```suggestion
#!/usr/bin/env bash
```
https://en.wikipedia.org/wiki/Shebang_(Unix)#Portability

**Review by AuHau — APPROVED** [review]
Generally LGTM 👍

Maybe you can add some description about this to the README?

**Pull request merged**